src/cirrus/lib/task.py

```diff
@@ -0,0 +1,208 @@
+                    'title': 'Source STAC Item',
+                    'rel': 'derived_from',
+                    'href': links[0],
+                    'type': 'application/json'
+                })
+
+    @abstractmethod
+    def process(self):
+        """Main task logic - virtua
+
+        Returns:
+            [type]: [description]
+        """
+        # download assets of interest, this will update self.items
+        #self.download_assets(['key1', 'key2'])
+        # do some stuff
+        #self.upload_assets(['key1', 'key2'])
+        pass
+
+    @property
+    def output_payload(self) -> Dict:
+        
+        # assemble return payload
+        payload = self.remaining_payload
+        payload.update({
+            'features': self.items,
+            'process': self.process_definition
+        })
+        return payload
+
+    @classmethod
+    def handler(cls, payload, **kwargs):
+        task = cls(payload, **kwargs)
+        try:
+            task.process()
+            return task.output_payload
+        except Exception as err:
+            task.logger.error(err, exc_info=True)
+            raise err
+        finally:
+            # remove work directory if not running locally
+            if task._tmpworkdir:
+                task.logger.debug('Removing work directory %s' % task._workdir)
+                rmtree(task._workdir)
+        
+    @classmethod
+    def parse_args(cls, args):
+        """ Parse CLI arguments """
+        dhf = argparse.ArgumentDefaultsHelpFormatter
+        parser0 = argparse.ArgumentParser(description=cls._description)
+
+        pparser = argparse.ArgumentParser(add_help=False)
+        pparser.add_argument('--version', help='Print version and exit', action='version', version=cls._version)
+        pparser.add_argument('--log', default=1, type=int,
+                                help='0:all, 1:debug, 2:info, 3:warning, 4:error, 5:critical')
+        subparsers = parser0.add_subparsers(dest='command')
+
+        # process subcommand
+        h = 'Locally process (development)'
+        parser = subparsers.add_parser('local', parents=[pparser], help=h, formatter_class=dhf)
+        parser.add_argument('filename', help='Full path of payload to process')
+        parser.add_argument('--workdir', help='Use this as work directory', default=None)
+        parser.add_argument('--save', help='Save output with provided filename', default=None)
+
+        # Cirrus process subcommand
+        h = 'Process Cirrus STAC Process Catalog'
+        parser = subparsers.add_parser('cirrus', parents=[pparser], help=h, formatter_class=dhf)
+        parser.add_argument('url', help='url (s3 or local) to Cirrus Process Payload')
+
+        # turn Namespace into dictionary
+        pargs = vars(parser0.parse_args(args))
+        # only keep keys that are not None
+        pargs = {k: v for k, v in pargs.items() if v is not None}
+
+        if pargs.get('command', None) is None:
+            parser.print_help()
```

Comment:
Should this be `pparser`?

Reply:
Actually should be parser0, the root parser.  But I moved out the parser build into it's own function separate from parsing the args, so users can now get the parser and add additional subcommands or args to it